components/tools/measure.js

```diff
@@ -3,377 +3,170 @@
-                    nearest = data;
-                }
+                container.select('rect')
+                    .attr('x', measure.xScale.value.range()[0])
+                    .attr('y', measure.yScale.value.range()[1])
+                    .attr('width', measure.xScale.value.range()[1])
+                    .attr('height', measure.yScale.value.range()[0]);
+
+                var g = fc.utilities.simpleDataJoin(container, 'measure', data);
+
+                var enter = g.enter();
+                enter.append('line')
+                    .attr('class', 'tangent');
+                enter.append('line')
+                    .attr('class', 'horizontal');
+                enter.append('line')
+                    .attr('class', 'vertical');
+                enter.append('text')
+                    .attr('class', 'horizontal');
+                enter.append('text')
+                    .attr('class', 'vertical');
+
+                g.select('line.tangent')
+                    .attr('x1', function(d) { return d.source.x; })
+                    .attr('y1', function(d) { return d.source.y; })
+                    .attr('x2', function(d) { return d.target.x; })
+                    .attr('y2', function(d) { return d.target.y; });
+
+                g.select('line.horizontal')
+                    .attr('x1', function(d) { return d.source.x; })
+                    .attr('y1', function(d) { return d.source.y; })
+                    .attr('x2', function(d) { return d.target.x; })
+                    .attr('y2', function(d) { return d.source.y; })
+                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; });
+
+                g.select('line.vertical')
+                    .attr('x1', function(d) { return d.target.x; })
+                    .attr('y1', function(d) { return d.target.y; })
+                    .attr('x2', function(d) { return d.target.x; })
+                    .attr('y2', function(d) { return d.source.y; })
+                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; });
+
+                var paddingValue = measure.padding.value.apply(this, arguments);
+
+                g.select('text.horizontal')
+                    .attr('x', function(d) { return d.source.x + (d.target.x - d.source.x) / 2; })
+                    .attr('y', function(d) { return d.source.y - paddingValue; })
+                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; })
+                    .text(measure.xLabel.value);
+
+                g.select('text.vertical')
+                    .attr('x', function(d) { return d.target.x + paddingValue; })
+                    .attr('y', function(d) { return d.source.y + (d.target.y - d.source.y) / 2; })
+                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; })
+                    .text(measure.yLabel.value);
+
```

Comment:
That should be handled by simpleDataJoin I think.

On Wed Jan 14 2015 at 4:04:35 PM James Leftley <notifications@github.com>
wrote:

> In components/tools/measure.js
> <https://github.com/ScottLogic/d3-financial-components/pull/244#discussion-diff-22946385>
> :
>
> > +                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; });
> > +
> > +                var paddingValue = measure.padding.value.apply(this, arguments);
> > +
> > +                g.select('text.horizontal')
> > +                    .attr('x', function(d) { return d.source.x + (d.target.x - d.source.x) / 2; })
> > +                    .attr('y', function(d) { return d.source.y - paddingValue; })
> > +                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; })
> > +                    .text(measure.xLabel.value);
> > +
> > +                g.select('text.vertical')
> > +                    .attr('x', function(d) { return d.target.x + paddingValue; })
> > +                    .attr('y', function(d) { return d.source.y + (d.target.y - d.source.y) / 2; })
> > +                    .style('visibility', function(d) { return d.state !== 'MEASURED' ? 'hidden' : 'visible'; })
> > +                    .text(measure.yLabel.value);
> > +
>
> I think g.exit().remove(); is missing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3-financial-components/pull/244/files#r22946385>
> .
>